### PR TITLE
Dh apmspii 874 draft implementation sync wrap

### PR DESCRIPTION
**Pull request by DominicH247**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: proxies/live/apiproxy/targets/sync-wrap.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (paruliansaragi) [inline proxies/live/apiproxy/targets/sync-wrap.xml:1]: Don't know if it makes a difference, but do we need these xml tags?

File: proxies/live/apiproxy/proxies/default.xml [hunk 2]
```diff
@@ -81,6 +81,12 @@
   <RouteRule name="NoRouteStatus">
     <Condition>(proxy.pathsuffix MatchesPath "/_status") and ((request.verb = "GET") or (request.verb = "HEAD"))</Condition>
   </RouteRule>
+  <RouteRule name="sync-wrap">
+    <Condition>
+      request.verb = "PATCH" AND (proxy.pathsuffix MatchesPath "/Patient/5900031454") AND request.header.prefer != "respond-async" AND request.header.x-sync-wrapped != "true"
```
Comment (DominicH247) [inline proxies/live/apiproxy/proxies/default.xml:86]: Temporarily hard coding NHS number

File: proxies/live/apiproxy/targets/sync-wrap.xml (hunk 1 above)
Comment (alexhawdon) [reply proxies/live/apiproxy/targets/sync-wrap.xml:1]: No strictly speaking - they are inferred if you leave them out

File: tests/user_restricted/data/pds_scenarios.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
...
...,"testing"],"period":{"start":"2021-01-22"},"postalCode":"HL12 6HL","use":"home"}],"birthDate":"2010-01-22","extension":[{"extension":[{"url":"language","valueCodeableConcept":{"coding":[{"code":"hy","display":"Armenian","system":"https://fhir.nhs.uk/R4/CodeSystem/UKCore-HumanLanguage","version":"1.0.0"}]}},{"url":"interpreterRequired","valueBoolean":False}],"url":"https://fhir.nhs.uk/R4/StructureDefinition/Extension-UKCore-NHSCommunication"}],"gender":"female","id":"5900038181","identifier":[{"extension":[{"url":"https://fhir.nhs.uk/R4/StructureDefinition/Extension-UKCore-NHSNumberVerificationStatus","valueCodeableConcept":{"coding":[{"code":"01","display":"Number present and verified","system":"https://fhir.nhs.uk/R4/CodeSystem/UKCore-NHSNumberVerificationStatus","version":"1.0.0"}]}}],"system":"https://fhir.nhs.uk/Id/nhs-number","value":"5900038181"}],"meta":{"security":[{"code":"U","display":"unrestricted","system":"https://www.hl7.org/fhir/valueset-security-labels.html"}],"versionId":"19"},"name":[{"family":"Bogan","given":["Testing"],"id":"7EB47E61","period":{"start":"2021-01-22"},"prefix":["Mr"],"suffix":["Bsc"],"use":"usual"}],"resourceType":"Patient"}},  # noqa: E231, E501
```
Comment (paruliansaragi) [inline tests/user_restricted/data/pds_scenarios.py:2]: This is more a problem of brittle tests i.e. do we need to have a "deep" compare of records?

**Review by paruliansaragi — DISMISSED** [review]

File: proxies/live/apiproxy/policies/AssignMessage.AddSyncWaitHeader.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<AssignMessage async="false" continueOnError="false" enabled="true" name="AssignMessage.AddSyncWaitHeader">
+  <Set>
+      <Headers>
+          <Header name="X-Sync-Wait">10</Header>
```
[the remark this answers is not in the document]
Comment (DominicH247) [reply proxies/live/apiproxy/policies/AssignMessage.AddSyncWaitHeader.xml:5]: 10s was agreed upon for now, then optimum to be determined through V&P testing. @alexhawdon correct me if i'm wrong

File: tests/user_restricted/data/pds_scenarios.py (hunk 3 above)
Comment (DominicH247) [reply tests/user_restricted/data/pds_scenarios.py:2]: this change was prior to your update to the patient retrieval test, and was needed to get the PR through the pipelines. I agree though this is a symptom of the brittle tests.

File: proxies/live/apiproxy/proxies/default.xml [hunk 5]
```diff
@@ -7,6 +7,12 @@
       <Step>
           <Name>AssignMessage.Copy.RequestHeaders</Name>
       </Step>
+      <Step>
```
[the remark this answers is not in the document]
Comment (DominicH247) [reply proxies/live/apiproxy/proxies/default.xml]: removed

File: proxies/live/apiproxy/policies/AssignMessage.Errors.RequestTimeout.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        <Name>pds.error.coding.code</Name>
+        <Value>UNABLE_TO_CALL_SERVICE</Value>
+    </AssignVariable>
+    <AssignVariable>
+        <Name>pds.error.coding.display</Name>
+        <Value>Unable to call Service</Value>
+    </AssignVariable>
+    <AssignVariable>
+        <Name>pds.error.diagnostics</Name>
+        <Template>For a SyncAsync wrapped request, the downstream domain processing has not completed within the configured timeout period.</Template>
```
Comment (alexhawdon) [inline proxies/live/apiproxy/policies/AssignMessage.Errors.RequestTimeout.xml:20]: This doesn't need to be `<template>`; can just be `<value>`

I can see the wording has come from https://nhsd-confluence.digital.nhs.uk/display/APM/2020-04-17+Spine+Error+Codes+II however it's not appropriate for public consumption - "SyncAsync wrapped..." is meaningless to our clients.

I'm sure I saw some wording, somewhere, about retrying the same `X-Request-ID`? I can't find it now though...but that's what the diagnostics/advice needs to be.

File: proxies/live/apiproxy/targets/sync-wrap.xml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        <Condition>(request.header.X-Sync-Wait LesserThan 0.25) OR (request.header.X-Sync-Wait GreaterThan 29) OR (request.header.X-Sync-Wait Is null)</Condition>
+      </Step>
+    </Request>
+  </PreFlow>
+  <LocalTargetConnection>
+    <APIProxy>sync-wrap-{{ APIGEE_ENVIRONMENT }}</APIProxy>
+    <ProxyEndpoint>default</ProxyEndpoint>
+    <Path>/sync-wrap/{{ SERVICE_BASE_PATH }}</Path>
+  </LocalTargetConnection>
+  <FaultRules>
```
Comment (alexhawdon) [inline proxies/live/apiproxy/targets/sync-wrap.xml:25]: Also need the `rate_limit_exceeded` Fault Rule from `ig3.xml` (minus the `AssignMessage.Mirror.*` steps)

File: proxies/live/apiproxy/policies/AssignMessage.Errors.RequestTimeout.xml (hunk 6 above)
Comment (paruliansaragi) [reply proxies/live/apiproxy/policies/AssignMessage.Errors.RequestTimeout.xml:20]: I can't find the `X-Request-ID` referenced anywhere in confluence although I did have it in my previous message. Can I make an exec decision on this and fix forward?

**Review by paruliansaragi — APPROVED** [review]

**Pull request merged**